server/src/test/java/org/elasticsearch/cluster/routing/UnassignedInfoTests.java

```diff
@@ -290,23 +307,168 @@ public void testFailedShard() {
+     * a `RESTART`-type shutdown, rather than the default global delay.
+     */
+    public void testRemainingDelayCalculationWhenNodeIsKnownToBeRestartingWithCustomDelay() throws Exception {
+        String lastNodeId = "bogusNodeId";
+        Map<String, SingleNodeShutdownMetadata> shutdowns = new HashMap<>();
+        SingleNodeShutdownMetadata shutdown = SingleNodeShutdownMetadata.builder()
+            .setNodeId(lastNodeId)
+            .setReason(this.getTestName())
+            .setStartedAtMillis(randomNonNegativeLong())
+            .setType(SingleNodeShutdownMetadata.Type.RESTART)
+            .setAllocationDelay(TimeValue.timeValueMinutes(1))
+            .build();
+        shutdowns.put(shutdown.getNodeId(), shutdown);
+
+        // Use a different index-level delay so this test will fail if that one gets used instead of the one from the shutdown metadata
+        checkRemainingDelayCalculation(lastNodeId, TimeValue.timeValueNanos(10), shutdowns, TimeValue.timeValueMinutes(1), true);
+    }
+
+    /**
+     * Verifies that the delay calculation uses the default delay value for nodes known to be restarting, because they are registered for
+     * a `RESTART`-type shutdown, rather than the default global delay.
      */
-    public void testRemainingDelayCalculation() throws Exception {
+    public void testRemainingDelayCalculationWhenNodeIsKnownToBeRestartingWithDefaultDelay() throws Exception {
+        String lastNodeId = "bogusNodeId";
+        Map<String, SingleNodeShutdownMetadata> shutdowns = new HashMap<>();
+
+        // Note that we do not explicitly configure the reallocation delay here.
+        SingleNodeShutdownMetadata shutdown = SingleNodeShutdownMetadata.builder()
+            .setNodeId(lastNodeId)
+            .setReason(this.getTestName())
+            .setStartedAtMillis(randomNonNegativeLong())
+            .setType(SingleNodeShutdownMetadata.Type.RESTART)
+            .build();
+        shutdowns.put(shutdown.getNodeId(), shutdown);
+
+        // Use a different index-level delay so this test will fail if that one gets used instead of the one from the shutdown metadata
+        checkRemainingDelayCalculation(
+            lastNodeId,
+            TimeValue.timeValueNanos(10),
+            shutdowns,
+            SingleNodeShutdownMetadata.DEFAULT_RESTART_SHARD_ALLOCATION_DELAY,
+            true
+        );
+    }
+
+    public void testRemainingDelayUsesIndexLevelDelayIfNodeWasNotRestartingWhenShardBecameUnassigned() throws Exception {
+        String lastNodeId = "bogusNodeId";
+        Map<String, SingleNodeShutdownMetadata> shutdowns = new HashMap<>();
+
+        // Generate a random time value - but don't use nanos as extremely small values of nanos can break assertion calculations
+        final TimeValue shutdownDelay = TimeValue.parseTimeValue(
+            randomTimeValue(2, 1000, "d", "h", "ms", "s", "m", "micros"),
```

Comment:
To avoid risking too many iterations on the `randomValueOtherThanMany` I propose:

```suggestion
            randomTimeValue(100, 1000, "d", "h", "ms", "s", "m", "micros"),
```